### ISSUE TITLE
minor: Chance to Hit displays wrong value/Hit chance is implemented incorrectly

### ISSUE THREAD

**Issue by coolrox95** [MelvorIdle/melvoridle.github.io#609]
**Describe the bug**
Dues to the way that hit chance is calculated, the values displayed in the combat page are incorrect. The display shows
```javascript
Math.floor(playerAccuracy)
```
when in reality it is
```javascript
Math.ceil(playerAccuracy)
```
**The issue with the implementation of accuracy**
In combat.js a player hitting an enemy first calculates the accuracy
```javascript
  let playerAccuracy;
  if (maximumAttackRoll < enemyDefenceRoll)
    playerAccuracy = ((0.5 * maximumAttackRoll) / enemyDefenceRoll) * 100;
  else
    playerAccuracy = (1 - (0.5 * enemyDefenceRoll) / maximumAttackRoll) * 100;  
```
Then a random integer from [0,99] (inclusive) is rolled:
```javascript
  let chanceToHit = Math.floor(Math.random() * 100);
```
Diamond luck applies and chooses the better chanceToHit value of two rolls,
Finally to determine if the player hits (if not through other means) the following inequality is used:
```javascript
  playerAccuracy > chanceToHit
```
This is equivalent to
```javascript
  playerAccuracy > Math.floor(Math.random()*100)
```
Which is equivalent to (see proof):
```javascript
  Math.ceil(playerAccuracy) > Math.random()*100
```
The proper way to roll for a percentage chance is:
```javascript
  chance > Math.random()*100
```
Therefore the hit chance of the player is actually
```javascript
 Math.ceil(playerAccuracy)
```
The function updatePlayerChanceToHit() contains the following code:
```javascript
  if (isInCombat && !newEnemyLoading) {
    let enemyDefenceRoll;
    if (attackStyle < 3) enemyDefenceRoll = combatData.enemy.maximumDefenceRoll;
    else if (attackStyle < 6)
      enemyDefenceRoll = combatData.enemy.maximumRangedDefenceRoll;
    else enemyDefenceRoll = combatData.enemy.maximumMagicDefenceRoll;
    if (maximumAttackRoll < enemyDefenceRoll)
      playerAccuracy = ((0.5 * maximumAttackRoll) / enemyDefenceRoll) * 100;
    else
      playerAccuracy = (1 - (0.5 * enemyDefenceRoll) / maximumAttackRoll) * 100;
    $('#combat-player-chance-to-hit').text(Math.floor(playerAccuracy) + '%');
  }
```
As you can see it is displaying an inaccurate value.
**Proving the above equivalence**
According to [this](https://janmr.com/blog/2009/09/useful-properties-of-the-floor-and-ceil-functions/) we know that the floor and ceiling functions have the following properties, where x is a real number, n is some integer:
1. `n > x` is equivalent to `n > floor(x)`
2. `x > n` is equivalent to `ceil(x) > n`

Consider the inequality
` a > floor(b)`
It follows from 2. that this expression is equivalent to
` ceil(a) > floor(b) `
Consider also:
` ceil(a) > b`
It follows from 1. that this expression is equivalent to
` ceil(a) > floor(b)`
Therefore the expressions `a > floor(b)` and `ceil(a) > b` are equivalent. 

**Expected behavior**
The combat page should either display the ceiling of ```playerAccuracy``` or the combat code should floor ```playerAccuracy```, or ```playerAccuracy``` should be implemented as the percentage hit chance.

**TLDR How to fix issue**
***Make the display show Math.ceil(accuracy)***
This keeps the mechanics the same, and fixes the display.
Change updateEnemyChanceToHit and updatePlayerChanceotHit so they set the text to the implemented value.
```javascript
$('#combat-enemy-chance-to-hit').text(Math.ceil(enemyAccuracy) + '%');
```
```javascript
$('#combat-player-chance-to-hit').text(Math.ceil(playerAccuracy) + '%');
```
***Fix the implementation of hitchance to match the display***
This nerfs player and enemy accuracy.
Requires flooring the `playerAccuracy` and `enemyAccuracy` in `attackEnemy()` and `attackPlayer()` after it is calculated,
and changing
```javascript
let chanceToHit = Math.floor(Math.random() * 100);
```
to
```javascript
let chanceToHit = Math.random() * 100;
```
Also need to change `diamondLuck` and `enemyChanceToHit` to use this formula.

***Make accuracy actually represent the % chance to hit***
Would nerf player and enemy accuracy slightly (less than the previous fix). This is possibly closer to original intent of code?
Requires changing the `chanceToHit`, `diamondLuck` and `enemyChanceToHit` calculations as per the previous fix.
Requires changing the display to show a decimal value
```javascript
$('#combat-player-chance-to-hit').text(playerAccuracy.toFixed(1) + '%');
```
You could also use [toPrecision](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toPrecision) if you want a number of significant digits, or Math.round() if you just want to display an integer.

**Side Note**
This issue may also be present in all the other cases where `Math.floor(Math.random()*100)` is used. (The flooring implementation is accurate when the chance is an integer value).
Also I spent too long writing this report.

**Browser**
Browser agnostic bug.

**Comment by MelvorIdle** [comment]
Thanks, fixed in upcoming update
